rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp

```diff
@@ -157,6 +157,9 @@ class SubscriptionIntraProcess : public SubscriptionIntraProcessBase
     msg_info.publisher_gid = {0, {0}};
     msg_info.from_intra_process = true;
 
+    if (!buffer_->has_data()) {
```

Comment:
I think that you want a `lock_guard` before this statement. I can imagine a segmentation fault as follows:
* Thread 1: `!buffer_->has_data()` evaluates to `false` so the thread skips over the if-statement
* Thread 2: consumes the data in `buffer_`
* Thread 1: Tries to consume the data that has already been consumed.

Reply:
Yeap, thanks for noting the issue!

Reply:
Solved in https://github.com/ros2/rclcpp/pull/1275/commits/3d0f8df8a86d93d362dc79cd7aaa95dcd9f9031b.